### PR TITLE
Custom resource schema modifications

### PR DESCRIPTION
**Pull request by danielrbradley**
Add a more understandable and unified approach to modifying existing parts of the schema.

Use a function to define a resource's schema–resource & types in both the schema and metadata. This is given any existing declaration which it is allowed to modify.

- If newly added types conflict with an existing type in the schema, we will raise an error.
- Allows modifying any of the types associated with the custom resource.
- Allow modification of both the schema and the metadata in one abstration.
- Allows creation of brand new resources - the existing resource will just be nil.

This allows the removal of the custom resource `Types`, `Schema` and `Meta` fields along with the associated `SchemaMixins`, `SchemaTypeMixins` and `MetaMixins` functions.

### CODE REVIEW

**Comment by danielrbradley** [comment]
> * I'm not sure why we need to discover all referenced types upfront, since most custom resources won't even use them. Could we offer `findTypesForResource` as a utility for custom resources instead, or even get away with just `TypeLookupFunc` for resources that know which types they want to access?

It felt like an easier API if you get passed values and you return values rather than callbacks. It avoids questions like "what happens if they call the callback lots of times" etc too. The types will only be searched if the custom resource implements the `SchemaModifications` function (which is only the Dashbaord right now). 

> * When two custom resources touch the same resource or type, the result depends on the order of iteration `customresources.ApplySchemaModifications`. It's sufficiently unlikely so it's ok, I think, but even better would be to keep track of modifications and warn if something is touched a second time.

Yes, good catch on the map iteration. Interestingly this is also the case for all the mixin functions too. I'll change them all to iterate by name for consistency at least.

> * Do we need `MetaTypes` when `SchemaModifications` can also add metadata?

Not strictly, but it's a simpler form to use if only adding and not modifying anything existing.

**Comment by thomas11** [comment]
> > * I'm not sure why we need to discover all referenced types upfront, since most custom resources won't even use them. Could we offer `findTypesForResource` as a utility for custom resources instead, or even get away with just `TypeLookupFunc` for resources that know which types they want to access?
> 
> It felt like an easier API if you get passed values and you return values rather than callbacks. It avoids questions like "what happens if they call the callback lots of times" etc too. The types will only be searched if the custom resource implements the `SchemaModifications` function (which is only the Dashbaord right now).

Makes sense!
 
> > * Do we need `MetaTypes` when `SchemaModifications` can also add metadata?
> 
> Not strictly, but it's a simpler form to use if only adding and not modifying anything existing.

In isolation, it's simpler, but having two ways to do the same thing is not simpler IMO (think of debugging).

**Comment by danielrbradley** [comment]
Rebased and squashed reworking following review .. and implemented the dashboard changes.

File: provider/pkg/resources/customresources/custom_portal_dashboard.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+			dashbordPartMetadata.Properties["settings"] = schema.PropertySpec{
+				// Remove the ref to the MarkdownPartSettings type and make it a loose object
+				TypeSpec: schema.TypeSpec{
+					Type:                 "object",
+					AdditionalProperties: &schema.TypeSpec{Ref: "pulumi.json#/Any"},
+				},
+				Description: "Dashboard part settings.",
+			}
+			resource.Types["azure-native:portal:DashboardPartMetadata"] = dashbordPartMetadata
+			// The response type is the same as the input type (there's no refs)
```
Comment (mikhailshilkov) [inline provider/pkg/resources/customresources/custom_portal_dashboard.go:62]: How do you know there are no refs in other markdown part properties? (in the future)

File: provider/pkg/resources/customresources/custom_portal_dashboard.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+			resource.Types["azure-native:portal:DashboardPartMetadataResponse"] = dashbordPartMetadata
+
+			// Mirror the changes in the metadata
```
Comment (mikhailshilkov) [inline provider/pkg/resources/customresources/custom_portal_dashboard.go:65]: Any possibility to reduce duplication and factor the changes into a reusable function?

File: provider/pkg/resources/customresources/custom_portal_dashboard.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+			dashbordPartMetadataMeta := markdownMetadataMeta
+			dashbordPartMetadataMeta.Properties["type"] = resources.AzureAPIProperty{
+				Type: "string",
+			}
+			dashbordPartMetadataMeta.Properties["settings"] = resources.AzureAPIProperty{
+				Type:                 "object",
+				AdditionalProperties: &resources.AzureAPIProperty{Ref: "pulumi.json#/Any"},
+			}
+			resource.MetaTypes["azure-native:portal:DashboardPartMetadata"] = dashbordPartMetadataMeta
+			resource.MetaTypes["azure-native:portal:DashboardPartMetadataResponse"] = dashbordPartMetadataMeta
```
Comment (mikhailshilkov) [inline provider/pkg/resources/customresources/custom_portal_dashboard.go:89]: Should we try to avoid hard-coding all these magic strings and build them instead?

File: provider/pkg/resources/customresources/custom_portal_dashboard.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+		tok: "azure-native:portal:Dashboard",
+		SchemaF: func(resource *ResourceDefinition) (*ResourceDefinition, error) {
+			if resource == nil {
+				// This should only happen if we're running with a namespace or version filter for testing, so we'll allow it to pass.
+				return nil, nil
+			}
+			// Add a new MetadataType to replace the MarkdownPartMetadata type
+			dashboardParts := resource.Types["azure-native:portal:DashboardParts"]
+			dashboardPartsMetadata := dashboardParts.Properties["metadata"]
+			dashboardPartsMetadata.Ref = "#/types/azure-native:portal:DashboardPartMetadata"
```
Comment (mikhailshilkov) [inline provider/pkg/resources/customresources/custom_portal_dashboard.go:35]: Should we assert on the current Ref value? In case the spec changes and doesn't point to MarkdownPart anymore?

**Review by mikhailshilkov — COMMENTED** [review]
This looks like a step in the right direction, thank you! Should we check for possibilities to rewrite some existing schema modifications in this style?

**Review by mjeffryes — COMMENTED** [review]
I agree with the general direction here, but there are a lot of details to sort out still. I suspect, it will be easier to reason about as a pure refactor, independent of restoring the dashboard customization. 

Would like to see this fully replace all the custom edits in `genMixins` as well.

File: provider/pkg/resources/customresources/customresources.go [hunk 5]
```diff
@@ -50,6 +56,81 @@ type CustomResource struct {
...
 
+type ResourceDefinition struct {
+	Resource     schema.ResourceSpec
+	Types        map[string]schema.ComplexTypeSpec
+	MetaResource AzureAPIResource
+	MetaTypes    map[string]AzureAPIType
+}
+
+// ApplySchemas applies custom schema modifications to the given package.
+// These modifications should never overlap with each other, but we apply in a deterministic order to ensure
```
[the remark this answers is not in the document]
Comment (danielrbradley) [reply provider/pkg/resources/customresources/customresources.go:67]: Yup, what Matt said - we _shouldn't_ ever have two custom resources modifying the same types, but in case we ever did, let's make sure it doesn't change every time we generate. I think trying to add validation that types aren't modified by two resources is tricky to write and not worth the time given how few custom resources we have (which don't overlap).

File: provider/pkg/resources/customresources/customresources.go [hunk 6]
```diff
@@ -50,6 +56,81 @@ type CustomResource struct {
...
+// that the end result of the modifications is consistent.
+func ApplySchemas(pkg *pschema.PackageSpec, meta *resources.AzureAPIMetadata) error {
```
[the remark this answers is not in the document]
Comment (danielrbradley) [reply provider/pkg/resources/customresources/customresources.go:69]: It's not only modifications. This function can also be used to create new resources and types. I think this will be clearer once the legacy functions are removed.

File: provider/pkg/resources/customresources/customresources.go [hunk 7]
```diff
@@ -50,6 +56,81 @@ type CustomResource struct {
...
+	}
+	return nil
+}
+
+func (r *CustomResource) ApplySchema(pkg *pschema.PackageSpec, meta *resources.AzureAPIMetadata) error {
+	if r.tok != "" && r.SchemaF != nil {
+		existingResource, ok := pkg.Resources[r.tok]
+		var originalResource *ResourceDefinition
+		types := map[string]pschema.ComplexTypeSpec{} // Hoist scope for easy lookup when checking if the type is new.
+		if ok {
```
[the remark this answers is not in the document]
Comment (danielrbradley) [reply provider/pkg/resources/customresources/customresources.go:86]: Exactly - the new schema function covers both modifying existing schemas as well as defining new schemas. Can rename the `ok` for clarity.

File: provider/pkg/resources/customresources/custom_portal_dashboard.go [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+// loosely typed collections to allow for arbitrary JSON objects.
+// 2. Overrides the 'metadata' property type of the portal:DashboardParts type to be DashboardPartMetadata.
+//
+// The override effectively removes the discriminated union and replaces it with a single loosely-typed structure.
+// We override both input and output (Response) types.
+func portalDashboard() *CustomResource {
+	return &CustomResource{
+		tok: "azure-native:portal:Dashboard",
+		SchemaF: func(resource *ResourceDefinition) (*ResourceDefinition, error) {
+			if resource == nil {
```
[the remark this answers is not in the document]
Comment (danielrbradley) [reply provider/pkg/resources/customresources/custom_portal_dashboard.go:28]: This is by design to allow this to be called for creating new resources too.

File: provider/pkg/resources/customresources/custom_portal_dashboard.go [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+				// This should only happen if we're running with a namespace or version filter for testing, so we'll allow it to pass.
+				return nil, nil
+			}
+			// Add a new MetadataType to replace the MarkdownPartMetadata type
```
[the remark this answers is not in the document]
Comment (danielrbradley) [reply provider/pkg/resources/customresources/custom_portal_dashboard.go:32]: MetadataType is the data used by the provider at runtime which should align to the schema types.

File: provider/pkg/resources/customresources/customresources.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (thomas11) [reply provider/pkg/resources/customresources/customresources.go:67]: Making sure the output is deterministic will just hide the bug, then. If anything, shouldn't we do the opposite and randomize so the issue becomes visible?

File: provider/pkg/resources/customresources/custom_portal_dashboard.go [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+			dashboardParts := resource.Types["azure-native:portal:DashboardParts"]
+			dashboardPartsMetadata := dashboardParts.Properties["metadata"]
+			dashboardPartsMetadata.Ref = "#/types/azure-native:portal:DashboardPartMetadata"
+			dashboardParts.Properties["metadata"] = dashboardPartsMetadata
+			resource.Types["azure-native:portal:DashboardParts"] = dashboardParts
+			// Do the same for the response type
```
[the remark this answers is not in the document]
Comment (danielrbradley) [reply provider/pkg/resources/customresources/custom_portal_dashboard.go:38]: Sure, wasn't clear how to do this well initially as each operation is on a different type - not keen on some heavy generics.

File: provider/pkg/resources/customresources/customresources.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (danielrbradley) [reply provider/pkg/resources/customresources/customresources.go:67]: I think the issue is how you identify what's causing the flip-flopping - it took us years to identify the similar issues in the examples generation.

File: provider/pkg/resources/customresources/custom_portal_dashboard.go (hunk 3 above)
Comment (danielrbradley) [reply provider/pkg/resources/customresources/custom_portal_dashboard.go:89]: The existing string builder functions were really difficult to read in order to see what we're actually creating in the schema. I think just writing out the strings is pretty clear to see what it will do.

File: provider/pkg/resources/customresources/custom_portal_dashboard.go [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+// The override does two things:
+// 1. Defines a new type DashboardPartMetadata with three properties: type, inputs, and settings. Inputs and settings are
+// loosely typed collections to allow for arbitrary JSON objects.
+// 2. Overrides the 'metadata' property type of the portal:DashboardParts type to be DashboardPartMetadata.
+//
+// The override effectively removes the discriminated union and replaces it with a single loosely-typed structure.
+// We override both input and output (Response) types.
+func portalDashboard() *CustomResource {
+	return &CustomResource{
+		tok: "azure-native:portal:Dashboard",
```
[the remark this answers is not in the document]
Comment (danielrbradley) [reply provider/pkg/resources/customresources/custom_portal_dashboard.go:26]: Not exactly sure how we iterate the set of available versions right now. I'd suggest we just leave this as-is for now – just fixing the default version.

File: provider/pkg/resources/customresources/customresources.go [hunk 12]
```diff
@@ -50,6 +56,81 @@ type CustomResource struct {
...
+				MetaResource: resourceMeta,
+				MetaTypes:    metaTypes,
+			}
+		}
+
+		customResource, err := r.SchemaF(originalResource)
+		if err != nil {
+			return fmt.Errorf("failed to apply custom resource schema modifications for %s: %w", r.tok, err)
+		}
+		if customResource == nil {
```
[the remark this answers is not in the document]
Comment (danielrbradley) [reply provider/pkg/resources/customresources/customresources.go:112]: A nil currently means no modification will be made. This is only really used currently when debugging generation and we don't recieve a source resource when we're expecting one. The ideal here would be to log a warning, but we don't have any kind of diagnostics framework implemented to allow this easily.

I'll just document the current behaviour and we can always adapt it later as needed.

File: provider/pkg/resources/customresources/custom_portal_dashboard.go [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+			dashboardParts.Properties["metadata"] = dashboardPartsMetadata
+			resource.Types["azure-native:portal:DashboardParts"] = dashboardParts
+			// Do the same for the response type
+			dashboardPartsResponse := resource.Types["azure-native:portal:DashboardPartsResponse"]
+			dashboardPartsMetadataResponse := dashboardPartsResponse.Properties["metadata"]
+			dashboardPartsMetadataResponse.Ref = "#/types/azure-native:portal:DashboardPartMetadata"
+			dashboardPartsResponse.Properties["metadata"] = dashboardPartsMetadataResponse
+			resource.Types["azure-native:portal:DashboardPartsResponse"] = dashboardPartsResponse
+
+			// Find the old MarkdownPartMetadata and derive the new DashboardPartMetadata from it
```
[the remark this answers is not in the document]
Comment (danielrbradley) [reply provider/pkg/resources/customresources/custom_portal_dashboard.go:45]: I'll switch approach so we're more similar to the existing implementation then.

File: provider/pkg/resources/customresources/customresources.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (danielrbradley) [reply provider/pkg/resources/customresources/customresources.go:69]: The new implementation mirrors the test approach of the existing generation code - which is covered largely via snapshot testing.

**Comment by danielrbradley** [comment]
I've rebased, addressed the minor comments and reworked the Portal schema to be much more similar to the existing code, with the addition of asserts over each type we're replacing to flag when there's changes upstream which affect the parts we're overwriting.

File: provider/pkg/resources/customresources/customresources.go [hunk 14]
```diff
@@ -30,11 +33,14 @@ type CustomResource struct {
...
-	// to our custom shape and behavior. Optional.
-	TypeOverrides map[string]schema.ComplexTypeSpec
 	// Resource schema. Optional, by default the schema is assumed to be included in Azure Open API specs.
 	Schema *schema.ResourceSpec
+	// Resource & types, schema & metadata modifications function.
+	// Optional, runs after main schema is generated allowing for modifications.
+	// The resource and types (both schema and metadata) returned will be written back into the schema and metadata.
+	// This can also include new types which were not present in the original schema.
+	// Any dangling, unreferenced types will be removed from the schema automatically.
+	SchemaF func(resource *ResourceDefinition) (*ResourceDefinition, error)
```
[the remark this answers is not in the document]
Comment (mikhailshilkov) [reply provider/pkg/resources/customresources/customresources.go:46]: Are you planning to move all custom resources to this? Looks like a surprising interface for pure custom resources.

File: provider/pkg/resources/customresources/customresources.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (mikhailshilkov) [reply provider/pkg/resources/customresources/customresources.go:79]: Given snapshot tests haven't changed in this PR, I don't see how they cover this code.

[the remark this answers is not in the document]
Comment (mikhailshilkov) [reply provider/pkg/resources/customresources/customresources.go:79]: I'd love for names to make sense for their current purpose, not for a future idea. Once the "legacy" functions are removed, we can rename.

File: provider/pkg/resources/customresources/customresources.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (mikhailshilkov) [reply provider/pkg/resources/customresources/customresources.go:77]: How was this comment addressed?

**Review by mikhailshilkov — COMMENTED** [review]
The changes since my last review look good. Still missing test coverage.

File: provider/pkg/resources/customresources/customresources.go [hunk 15]
```diff
@@ -102,6 +191,7 @@ func BuildCustomResources(env *azureEnv.Environment,
 		portalDashboard(),
 		customWebApp,
 		customWebAppSlot,
+		portalDashboard(),
```
Comment (mikhailshilkov) [inline provider/pkg/resources/customresources/customresources.go:194]: Just noticed you added it for the second time here

File: provider/pkg/resources/customresources/customresources.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (danielrbradley) [reply provider/pkg/resources/customresources/customresources.go:46]: Yes, because it's then a single function to define all the parts of the schema rather than various separate properties which are amalgamated into a single resource.

File: provider/pkg/resources/customresources/customresources.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (danielrbradley) [reply provider/pkg/resources/customresources/customresources.go:85]: IMO this makes sense for what it can currently do, even if it's not yet being used for it. The only confusing part is that the legacy code exists at all. I've renamed `Schema` to `LegacySchema` and `SchemaF` to `Schema` for better forward naming.

[the remark this answers is not in the document]
Comment (danielrbradley) [reply provider/pkg/resources/customresources/customresources.go:85]: I've pushed the snapshot test

File: provider/pkg/resources/customresources/customresources.go (hunk 15 above)
Comment (danielrbradley) [reply provider/pkg/resources/customresources/customresources.go]: 👍 Fixed

File: provider/pkg/gen/gen_dashboard_test.go [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+
+	providers, _, err := openapi.ReadAzureProviders(path.Join(rootDir, "azure-rest-api-specs"), "Portal", "2020-09-01-preview")
+	if err != nil {
+		t.Fatal(err)
+	}
+	generationResult, err := PulumiSchema(rootDir, providers, versioningStub{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	snaps.MatchSnapshot(t, generationResult.Schema)
```
Comment (mikhailshilkov) [inline provider/pkg/gen/gen_dashboard_test.go:34]: In my opinion, snapshot tests are no substitute to proper unit tests. It was okay to add them post-factum for all the existing code that we had no tests for, but I suggest we don't use them as an excuse not to write tests for net-new non-trivial functionality.

Your current PR is a nice illustration of my case: as far as I can tell, your snapshot is actually not testing anything at all. The only `portal` resource that the snapshot contains is `azure-native:portal/v20200901preview:Dashboard`, which is not processed by the custom resource you implemented, so it just passes `nil` to the `Schema` method, and the method exists immediately. Yet, the "test" provides an impression that we are testing something, and it's hard to tell otherwise without looking into the details.

@thomas11 @mjeffryes I'd love to get your opinion on this testing approach and whether it matches mine or not.

File: provider/pkg/resources/customresources/custom_portal_dashboard.go [hunk 17]
```diff
@@ -119,6 +168,39 @@ func dashboardPartMetadataType() schema.ComplexTypeSpec {
...
+						Type: "object",
+						Ref:  fmt.Sprintf("#/types/azure-native:portal:MarkdownPartMetadataSettings%s", suffix),
+					},
+					Description: "Markdown part settings.",
+				},
+				"type": {
+					TypeSpec: schema.TypeSpec{
+						Type: "string",
+					},
+					Description: "The dashboard part metadata type.\nExpected value is 'Extension/HubsExtension/PartType/MarkdownPart'.",
```
Comment (mikhailshilkov) [inline provider/pkg/resources/customresources/custom_portal_dashboard.go:194]: On the second pass, I'm not sure I like the strictness of this. We will have to re-write this code in case upstream changes the description of a type that we are throwing away, right? That seems unnecessary fragile?

File: provider/pkg/resources/customresources/custom_portal_dashboard.go [hunk 18]
```diff
@@ -31,10 +32,32 @@ func portalDashboard() *CustomResource {
...
 			resource.Types["azure-native:portal:DashboardParts"] = dashboardPartsType("")
+			existingDashboardPartsTypeResponse := resource.Types["azure-native:portal:DashboardPartsResponse"]
+			if !reflect.DeepEqual(existingDashboardPartsTypeResponse, expectedDashboardPartsType("Response")) {
+				return nil, fmt.Errorf("unexpected azure-native:portal:DashboardPartsResponse type: %#v", existingDashboardPartsTypeResponse)
+			}
 			resource.Types["azure-native:portal:DashboardPartsResponse"] = dashboardPartsType("Response")
 
+			existingMarkdownPartMetadataType := resource.Types["azure-native:portal:MarkdownPartMetadata"]
+			if !reflect.DeepEqual(existingMarkdownPartMetadataType, expectedMarkdownPartMetadataType("")) {
+				return nil, fmt.Errorf("unexpected azure-native:portal:MarkdownPartMetadata type: %#v", existingMarkdownPartMetadataType)
```
Comment (mikhailshilkov) [inline provider/pkg/resources/customresources/custom_portal_dashboard.go:48]: Should we provide guidance to future maintainers about what they need to do in this situation?

File: provider/pkg/gen/gen_dashboard_test.go (hunk 16 above)
Comment (danielrbradley) [reply provider/pkg/gen/gen_dashboard_test.go:34]: Yes, it looks like the snapshot currently misses generating the default resource, I'll take a look at fixing that when I get opportunity.

@mikhailshilkov Perhaps if you're envisioning something specific with the unit tests that you think is more effective, you could contribute it?

File: provider/pkg/resources/customresources/custom_portal_dashboard.go (hunk 17 above)
Comment (danielrbradley) [reply provider/pkg/resources/customresources/custom_portal_dashboard.go:194]: This is why the original design I proposed attempted to derrive the modified schema from the original schema in order to be more flexible, but the feedback on that was that it was too complex & fragile. The frequency of updates hasn't been too high previously so we can always refine these asserts if needed later.

File: provider/pkg/resources/customresources/custom_portal_dashboard.go (hunk 18 above)
Comment (danielrbradley) [reply provider/pkg/resources/customresources/custom_portal_dashboard.go:48]: Sure, can refine these message further, though having any messages is already a big improvement over the current failure mode.

File: provider/pkg/gen/gen_dashboard_test.go (hunk 16 above)
Comment (thomas11) [reply provider/pkg/gen/gen_dashboard_test.go:34]: Apart from the snaps vs unit tests question, a test that passes without asserting anything is a bug. Can we add some assertions here to prevent that? Along the lines of "`generationResult` contains foo".

In general, I'm all for unit tests over snap-based tests, but for things the portal resource where you have a bunch of opaque JSON, snaps can still a good fit.

Comment (mikhailshilkov) [reply provider/pkg/gen/gen_dashboard_test.go]: How can we test anything except the happy path with snapshot tests? I guess we miss a mechanism to amend snapshots to cause errors/edge cases and assert the outcomes?

Comment (thomas11) [reply provider/pkg/gen/gen_dashboard_test.go]: My understanding is that snapshot tests, as we're using them, are strictly for the happy path.

File: provider/pkg/gen/schema_test.go [hunk 19]
```diff
@@ -33,17 +33,29 @@ func TestTypeAliasFormatting(t *testing.T) {
 var _ Versioning = (*versioningStub)(nil)
 
 type versioningStub struct {
+	shouldIncline   func(provider string, version string, typeName, token string) bool
```
Comment (thomas11) [inline provider/pkg/gen/schema_test.go:36]: "Incline"?

Comment (danielrbradley) [reply provider/pkg/gen/schema_test.go:36]: Oops, yes - should be include. Ended up taking out the code which calls this mock. Will rename as I rebase 👍

**Comment by mikhailshilkov** [comment]
Note that the "Dashboard Fix" section of the description was not correct anymore, so I deleted it
